# Blast-Off
A pipeline for our research workflow.

Hello! If you are reading this, you are most likely a part of our CRSP research group. This program with made for the purpose of automating
a portion of our workflow and make it easier to annotate the staph genes. If you are not a part of our group, below is the abstract for our 
research project.

Abstract: Staphylococci are gram-positive aerobic microorganisms. Most strains are persistent and highly diverse bacteria that inhabits
the mucous membranes of those who are infected. About 30% of people are unaware that they carriers of the bacterium, as it can inhabit a 
person’s body without any showing symptoms. However, the bacterium is capable of proliferating through cuts in the skin or the weakening 
of the infected individual’s immune system. When given ground, the infection can cause symptoms, such as the growth of pimples and boils 
to large abscesses and the destruction of skin tissue. The diversity of the bacterium comes from its adaptability and resistance towards 
certain antibiotics. Due to the adaptable nature of the infection, hospitals are unable to sufficiently treat their patients unless 
adequate information about the strain is known. However, with metagenomic analysis, the direct genetic analysis of genomes contained
with an environmental sample, more data on the strains can be collected. Thorough this technique, an adequate selection of antibiotics 
can be efficiently made to treat the patients. Still, the Multidrug-resistance Genes in Staph species, which are a strain of the infection 
that has become resistant to the use of multiple antibiotics, are a cause for concern. Owing to the diversity of the bacterium, 
it is difficult to gather information on the various strains due to their distinct resistances to antibiotics. Using results from 
the Blast analysis of genes from the staph strain, USA300, a strain of community-associated methicillin-resistant Staphylococcus aureus 
(MRSA), compared across 161 other genomes, we’re able to properly annotate these genes. Through manual annotation, we’re able to give 
them informative names and offer proper information about the genes.  

